src/hardware/serialport/serialmouse.cpp

```diff
@@ -0,0 +1,453 @@
+        break;
+    case MouseType::LOGITECH:
+        packet[0]  = 'M';
+        packet[1]  = '3';
+        packet_len = 2;
+        break;
+    case MouseType::WHEEL:
+        packet[0]  = 'M';
+        packet[1]  = 'Z';
+        packet[2]  = '@'; // for some reason 86Box sends more than 'MZ'
+        packet[3]  = 0;
+        packet[4]  = 0;
+        packet[5]  = 0;
+        packet_len = 6;
+        break;
+    case MouseType::MOUSE_SYSTEMS:
+        packet[0]  = 'H';
+        packet_len = 1;
+        break;
+    default:
+        unimplemented();
+        break;
+    }
+
+    // send packet
+    xmit_idx = 0;
+    setEvent(SERIAL_RX_EVENT, bytetime / smooth_div);
+}
+
+void CSerialMouse::startPacketData(bool extended) {
+    if (!mouse_port_valid) return;
+
+    if (mouse_type == MouseType::MICROSOFT ||
+        mouse_type == MouseType::LOGITECH  ||
+        mouse_type == MouseType::WHEEL) {
+        //          -- -- -- -- -- -- -- --
+        // Byte 0:   X  1 LB RB Y7 Y6 X7 X6
+        // Byte 1:   X  0 X5 X4 X3 X2 X1 X0
+        // Byte 2:   X  0 Y5 Y4 Y3 Y2 Y1 Y0
+        // Byte 3:   X  0 MB 00 W3 W2 W1 W0  - only sent if needed
+
+        // Do NOT set bit 7. It confuses CTMOUSE.EXE (CuteMouse) serial support.
+        // Leaving it clear is the only way to make mouse movement possible.
+        // Microsoft Windows on the other hand doesn't care if bit 7 is set.
+
+        uint8_t dx = static_cast<uint8_t>(std::clamp(mouse_delta_x, -0x80, 0x7f));
+        uint8_t dy = static_cast<uint8_t>(std::clamp(mouse_delta_y, -0x80, 0x7f));
+        uint8_t bt = mouse_has_3rd_button ? (mouse_buttons & 7) : (mouse_buttons & 3);
+
+        packet[0]  = static_cast<uint8_t>(0x40 | ((bt & 1) << 5) | ((bt & 2) << 3) | (((dy >> 6) & 3) << 2) | ((dx >> 6) & 3));
+        packet[1]  = static_cast<uint8_t>(0x00 | (dx & 0x3f));
+        packet[2]  = static_cast<uint8_t>(0x00 | (dy & 0x3f));
+        if (extended) {
+            uint8_t dw = std::clamp(mouse_delta_w, -0x10, 0x0f) & 0x0f;
+            packet[3]  = static_cast<uint8_t>(((bt & 4) ? 0x20 : 0) | dw);
+            packet_len = 4;
+        } else {
+            packet_len = 3;           
+        }
+        xmit_2part = false;
+
+    } else if (mouse_type == MouseType::MOUSE_SYSTEMS) {
+        //          -- -- -- -- -- -- -- --
+        // Byte 0:   1  0  0  0  0 LB MB RB
+        // Byte 1:  X7 X6 X5 X4 X3 X2 X1 X0
+        // Byte 2:  Y7 Y6 Y5 Y4 Y3 Y2 Y1 Y0
+
+        uint8_t dx = static_cast<uint8_t>(std::clamp(mouse_delta_x,  -0x80, 0x7f));
+        uint8_t dy = static_cast<uint8_t>(std::clamp(-mouse_delta_y, -0x80, 0x7f));
+        uint8_t bt = mouse_has_3rd_button ? ((~mouse_buttons) & 7) : ((~mouse_buttons) & 3);
+
+        packet[0]  = static_cast<uint8_t>(0x80 | ((bt & 1) << 2) | ((bt & 2) >> 1) | ((bt & 4) >> 1));
+        packet[1]  = dx;
+        packet[2]  = dy;
+        packet_len = 3;
+        xmit_2part = true; // next part contains mouse movement since the start of the 1st part
```

Comment:
I think this code would massively benefit from @kcgen's excellent new `bit_view` helper class specifically designed for such bit-twiddling stuff.

Reply:
The problem with bit_view is, that I would need to declare several unions, use each of them only once, and afterwards copy all the bytes (as you can't use the bit view for array element). The code to construct data packages would have grown in size several times - so I prefer to keep it the current way.

Reply:
Up to you, but I still think it would improve readability greatly overall. Copying a few bytes is a complete non-issue if you're worried about performance. Maybe you'd see a measurable difference (still tiny) if you copied a few Ks worth of data, but not a few bytes (and the compiler would optimise it away, most likely).